### PR TITLE
New script to automate IanaTimeZoneMappings.xml fix

### LINKED ISSUE #2241: Script to automate IanaTimeZoneMappings.xml fix
We need to publish a script that can be used to fix the problematic IanaTimeZoneMappings.xml. The goal is to make it easier for customers to apply the fix.

The script which is provided in the KB works as expected but doesn’t support to process multiple servers at once.

https://support.microsoft.com/en-us/topic/time-zone-exception-occurs-after-installing-exchange-server-november-2024-su-version-1-or-version-2-851b3005-6d39-49a9-a6b5-5b4bb42a606f

### PR DESCRIPTION
**Pull request by lusassl-msft**
**Issue:**
This script automates the fix for the `IanaTimeZoneMappings.xml` file, which is corrupt since the Exchange Server November 2024 Security Update (both versions).

More information: https://support.microsoft.com/en-us/topic/time-zone-exception-occurs-after-installing-exchange-server-november-2024-su-version-1-or-version-2-851b3005-6d39-49a9-a6b5-5b4bb42a606f

Resolve #2241 

**Reason:**
Make it easier for customers to apply the fix

**Fix:**
Script that automates the fix. It can be executed against one or multiple Exchange servers.

**Validation:**
Lab / Test team (pending)

### CODE REVIEW

File: Admin/Remove-DuplicateEntriesFromIanaMappings.ps1 [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+.EXAMPLE
+    PS C:\> Get-ExchangeServer | .\Remove-DuplicateEntriesFromIanaMappings.ps1
+.EXAMPLE
+    PS C:\> .\Remove-DuplicateEntriesFromIanaMappings.ps1 -ComputerName exch1.contoso.com -RestartServices $true
+#>
+
+[CmdletBinding(DefaultParameterSetName = "Default", SupportsShouldProcess, ConfirmImpact = 'High')]
+param (
+    [Parameter(Mandatory = $false, ValueFromPipeline = $true, ParameterSetName = "Default")]
+    [string[]]$ComputerName = $env:COMPUTERNAME,
```
Comment (lusassl-msft) [inline Admin/Remove-DuplicateEntriesFromIanaMappings.ps1:34]: Note to myself: Rename parameter to `Server`

File: Admin/Remove-DuplicateEntriesFromIanaMappings.ps1 [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        Write-Host "The script needs to be executed in elevated mode. Start the PowerShell as an administrator." -ForegroundColor Yellow
+        exit
+    }
+
+    foreach ($srv in $Server) {
+        # Check if the target server is online / reachable to us, we use the Get-ExSetupFileVersionInfo custom function to do this
+        $exchangeFileVersionInfo = Get-ExSetupFileVersionInfo -Server $srv
+
+        if ($null -ne $exchangeFileVersionInfo -and
+            (-not([System.String]::IsNullOrEmpty($exchangeFileVersionInfo)))) {
```
Comment (dpaulson45) [inline Admin/Remove-DuplicateEntriesFromIanaMappings.ps1:195]: Don't think both of these are needed.

File: Admin/Remove-DuplicateEntriesFromIanaMappings.ps1 [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+            try {
+                $writeProgressParams.Activity = $activityBase + " Creating backup $mappingBackupFilePath"
+                Write-Progress @writeProgressParams
+
+                # If duplicate entries were detected, create a backup of the file before modifying it
+                Copy-Item -Path $mappingFilePath -Destination $mappingBackupFilePath
+                Write-Host "[+] Backup created: $mappingBackupFilePath"
+            } catch {
+                Write-Host "[+] Failed to create backup file. Inner Exception: $_" -ForegroundColor Red
+                Write-Progress @writeProgressParams -Completed
```
Comment (dpaulson45) [inline Admin/Remove-DuplicateEntriesFromIanaMappings.ps1:125]: Recommend wrapping everything in a `try` `finally` block that handles the `Write-Progress` inside the `finally` block you will always do `Write-Progress @writeProgressParams -Completed` so you would just need to do a `return` when you don't want to continue with the code.

File: Admin/Remove-DuplicateEntriesFromIanaMappings.ps1 [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+                    $ianaMappingXml.Save($mappingFilePath)
+
+                    # Restart services if the -RestartServices $true was used when running the script - otherwise do nothing
+                    if ($PerformServiceRestart) {
+                        Write-Host "[+] Restart services: W3SVC, WAS & MSExchangeTransport"
+
+                        $writeProgressParams.Activity = $activityBase + " Restarting services W3SVC, WAS & MSExchangeTransport"
+                        Write-Progress @writeProgressParams
+
+                        Restart-Service -Name W3SVC, WAS, MSExchangeTransport -Force
```
Comment (dpaulson45) [inline Admin/Remove-DuplicateEntriesFromIanaMappings.ps1:171]: Can this throw an exception?

File: Admin/Remove-DuplicateEntriesFromIanaMappings.ps1 [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+    Write-Verbose "PowerShell version: $($PSVersionTable.PSVersion)"
+
+    $scriptBlock = {
+        param(
+            $PerformServiceRestart,
+            $LocalComputerName
+        )
+
+        . $PSScriptRoot\..\Diagnostics\HealthChecker\Analyzer\Test-IanaTimeZoneMapping.ps1
```
Comment (dpaulson45) [inline Admin/Remove-DuplicateEntriesFromIanaMappings.ps1:60]: Move to `Shared` as discussed in previous PR.

**Review by dpaulson45 — CHANGES_REQUESTED** [review]

File: Admin/Remove-DuplicateEntriesFromIanaMappings.ps1 (hunk 2 above)
Comment (lusassl-msft) [reply Admin/Remove-DuplicateEntriesFromIanaMappings.ps1:195]: Addressed with next commit.

File: Admin/Remove-DuplicateEntriesFromIanaMappings.ps1 (hunk 4 above)
Comment (lusassl-msft) [reply Admin/Remove-DuplicateEntriesFromIanaMappings.ps1:171]: There is a chance. Will place this into its own try/catch block just to make sure. Resolved with next commit.

File: Admin/Remove-DuplicateEntriesFromIanaMappings.ps1 (hunk 5 above)
Comment (lusassl-msft) [reply Admin/Remove-DuplicateEntriesFromIanaMappings.ps1:60]: Addressed with next iteration.

File: Admin/Remove-DuplicateEntriesFromIanaMappings.ps1 (hunk 3 above)
Comment (lusassl-msft) [reply Admin/Remove-DuplicateEntriesFromIanaMappings.ps1:125]: Good idea. Will add this with the next iteration.

**Review by dpaulson45 — APPROVED** [review]

**Pull request merged**